spec/controllers/line_items_controller_spec.rb

```diff
@@ -0,0 +1,165 @@
+require 'spec_helper'
+
+describe LineItemsController do
+  let(:user) { create(:user) }
+  let(:distributor) { create(:distributor_enterprise) }
+  let(:order_cycle) { create(:simple_order_cycle) }
+
+  context "listing bought items" do
+    let!(:completed_order) do
+      order = create(:completed_order_with_totals, user: user, distributor: distributor, order_cycle: order_cycle)
+      while !order.completed? do break unless order.next! end
+      order
+    end
+
+    before do
+      controller.stub spree_current_user: user
+      controller.stub current_order_cycle: order_cycle
+      controller.stub current_distributor: distributor
+    end
+
+    it "lists items bought by the user from the same shop in the same order_cycle" do
+      get :index, { format: :json }
+      expect(response.status).to eq 200
+      json_response = JSON.parse(response.body)
+      expect(json_response.length).to eq completed_order.line_items(:reload).count
+      expect(json_response[0]['id']).to eq completed_order.line_items.first.id
+    end
+  end
+
+  describe "destroying a line item on a completed order" do
+    let(:item) do
+      order = create(:completed_order_with_totals)
+      item = create(:line_item, order: order)
+      while !order.completed? do break unless order.next! end
+      item
+    end
+
+    before { controller.stub spree_current_user: item.order.user }
+
+    context "without a line item id" do
+      it "fails and raises an error" do
+        expect { delete :destroy }.to raise_error
+      end
+    end
+
+    context "with a line item id" do
+      let(:params) { { format: :json, id: item } }
+
+      context "without a user" do
+        it "denies deletion" do
+          delete :destroy, params
+          expect(response.status).to eq 403
+          expect { item.reload }.to_not raise_error
+        end
+      end
+
+      context "where the item's order is associated with the current user" do
+        before { item.order.update_attributes(user_id: user.id) }
+
+        context "without an order cycle" do
+          it "denies deletion" do
+            delete :destroy, params
+            expect(response.status).to eq 403
+            expect { item.reload }.to_not raise_error
+          end
+        end
+
+        context "with an order cycle" do
+          before { item.order.update_attributes(order_cycle_id: order_cycle.id) }
+
+          context "without a distributor" do
+            it "denies deletion" do
+              delete :destroy, params
+              expect(response.status).to eq 403
+              expect { item.reload }.to_not raise_error
```

Comment:
Good question! It seems to be checking that this line `item = Spree::LineItem.find(params[:id])` in the destroy method on the controller doesn't raise an error. Which it should never do, really. I think this should probably be using rescued and return a 404...